Add guide for setting up a free home server (#597)
matrix-org/matrix.org: gatsby/content/docs/free-small-matrix-server.mdx

---
title: 'Free Small Matrix Server'
author: Paul Tötterman <paul.totterman@iki.fi>
layout: post
section: guides
license: CC-BY-NC-SA ?
slug: free-small-matrix-server
---

*Canonical version of this article at <https://gitlab.com/ptman/matrix-docs/tree/master/free-matrix-server>*

# Free Small Matrix Server

Ingredients:

- Domain
- Server
- Software
- Configuration

## Free domain name

Get one from [Freenom] for free or any other place you wish. Freenom only gives

> free domains for a year. A Matrix server cannot be migrated (yet?) from one

Reviewer: ```suggestion
free domains for a year. A Matrix server cannot be migrated from one
```

yep :P

domain to another, so you might want to get a longer-lived one.

- https://tld-list.com/ - find cheap domains, look at renewal fees also
- https://www.namecheap.com/promos/amazing98s/ - cheap first year

> You also need to specify subdomains (which is why most dynamic dns services

Reviewer: Which subdomains?

Edit: Oh, riot./matrix. ?

Author: so, it's ok as they're mentioned later?

If you choose to, register on Oracle Cloud (requires credit card for
verification) and create a free VM:

1. Create a VM instance
2. Give it a name
3. OS/image: Canonical Ubuntu 18.04 Minimal
4. Show Shape, Network and Storage Options -> Assign a public IP address
5. Add SSH key ([generate one][sshkey] using `ssh-keygen` if you don't have one)
6. Show advanced options. I removed monitoring, since I'll remove the monitoring
   agent later.

View resources and make sure the instance and boot volume are "Always Free" if
you don't intent to pay for them.

Make sure you can ssh to the server (using the ssh key you generated/picked)
using the user 'ubuntu' and can use `sudo`.

[sshkey]: https://docs.oracle.com/en/cloud/iaas/compute-iaas-cloud/stcsg/generating-ssh-key-pair.html

## Open firewall

### Oracle Cloud Security lists

View resources -> Instances -> Select instance -> Virtual Cloud Network ->
Public Subnet -> Security Lists -> Default -> Ingress

Open incoming for CIDR 0.0.0.0/0:
- 22/tcp for SSH (should be open already)
- 80/tcp for HTTP
- 443/tcp for HTTPS
- 8448/tcp for Matrix federation
- 3478/tcp, 5349/tcp, 3478/udp, 5349/udp, 49152-49172/udp for TURN/STUN (you can

>   omit this if you don't run coturn for audio and video calls)

Reviewer: New users will likely not know what this means. I might reword to:

```suggestion
  omit this if you don't wish to support 1-to-1 audio and video calls)
```

Author: I've been thinking about this wording, since it's not actually needed for calls, but usually helps if there are firewalls or NATs

Author: hope you like this wording better

Reviewer 2: While it isn't *strictly required* for calls, it is basically required if you want calls to work reliably so personally I prefer anoa's wording.

Author: Ok, I just don't want to give the wrong impression of it being needed. Lots of places that don't care about calls, just need an alternative to IRC/Slack. I tried to cover this in one place

### Ubuntu iptables

The Oracle Cloud Ubuntu images come with somewhat restrictive iptables rules by
default. Docker manages the instance firewall and we have the Oracle Cloud
firewall in front, so let's remove the current firewall to avoid trouble:

```sh
apt purge netfilter-persistent iptables-persistent
```

## Remove useless stuff (optional)

Oracle cloud includes a somewhat heavy monitoring daemon. We have better use for

> that memory since current versions of synapse, the Matrix homeserver, can be

Reviewer: ```suggestion
that memory since current versions of Synapse, the Matrix homeserver, can be
```

- CNAME record `riot.$domain` pointing to `$domain`

## Use matrix-docker-ansible-deploy

Follow the [guide], with some tweaks:

[guide]: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/master/docs/README.md

### Guide in a nutshell, TL;DR

```sh
git clone https://github.com/spantaleev/matrix-docker-ansible-deploy/
cd matrix-docker-ansible-deploy
mkdir inventory/host_vars/matrix.$domain
cp examples/host-vars inventory/host_vars/matrix.$domain/vars.yaml
cp examples/hosts inventory/hosts
$EDITOR inventory/hosts
$EDITOR inventory/host_vars/matrix.$domain/vars.yaml
ansible-playbook -i inventory/hosts setup.yml --tags=setup-all
ansible-playbook -i inventory/hosts setup.yml --tags=start
ansible-playbook -i inventory/hosts setup.yml --tags=self-check
ansible-playbook -i inventory/hosts setup.yml -e username=$user -e password=$pass -e admin=yes --tags=register-user
```

### Serve the base domain as well (unless you already have something serving it)

```yaml
matrix_nginx_proxy_base_domain_serving_enabled: true
```

### Disable all extras because of RAM

```yaml
matrix_dimension_enabled: false
matrix_mxisd_enabled: false
matrix_mailer_enabled: false

> matrix_coturn_enabled: true # disable to save more ram, affects audio/video

Reviewer: ```suggestion
matrix_coturn_enabled: true # disable to save more ram, enables 1-to-1 audio/video support
```

```

> ### Configure the public IP address for coturn

Reviewer: ```suggestion
### Configure the public IP address for coturn

Optional if you want to support 1-to-1 audio/video calls.
```

```yaml
matrix_coturn_turn_external_ip_address: $instance_external_ip_address
```

### Limit rooms because of limited RAM

```yaml
matrix_synapse_configuration_extension_yaml: |
  limit_remote_rooms:
    enabled: true

>     complexity: 1.0

Reviewer: This is quite a low value. it's fine to leave this, but you should explain what it will affect (joining large rooms) and that it can be tweaked later if necessary.

Author: are there any guidelines for suitable values?

```

## Done!

Point your browser to https://riot.$domain or use [another
client](https://matrix.org/clients).

## Maintenance

Remember to keep your VM up to date!

```sh
apt update
apt upgrade
reboot # e.g. kernel,dbus,systemd updates
```

To keep the matrix services upgraded, start by reading the [maintenance docs],
but really, take a brief look at all the documentation.

[maintenance docs]: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/master/docs/maintenance-upgrading-services.md

Come join [#synapse:matrix.org](https://matrix.to/#/#synapse:matrix.org) to
discuss homeservers. Or say hi to

> [@ptman:ptman.name](https://matrix.to/#/@ptman:ptman.name)

Reviewer: ```suggestion
[@ptman:ptman.name](https://matrix.to/#/@ptman:ptman.name), who wrote this guide.
```

records:

- A record `$domain` pointing to `$instance_external_ip_address`
- CNAME record `matrix.$domain` pointing to `$domain`
- CNAME record `riot.$domain` pointing to `$domain`

## Use matrix-docker-ansible-deploy

Follow the [guide], with some tweaks:

[guide]: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/master/docs/README.md

### Guide in a nutshell, TL;DR

```sh
git clone https://github.com/spantaleev/matrix-docker-ansible-deploy/
cd matrix-docker-ansible-deploy
mkdir inventory/host_vars/matrix.$domain
cp examples/host-vars inventory/host_vars/matrix.$domain/vars.yaml
cp examples/hosts inventory/hosts
$EDITOR inventory/hosts
$EDITOR inventory/host_vars/matrix.$domain/vars.yaml
ansible-playbook -i inventory/hosts setup.yml --tags=setup-all
ansible-playbook -i inventory/hosts setup.yml --tags=start
ansible-playbook -i inventory/hosts setup.yml --tags=self-check
ansible-playbook -i inventory/hosts setup.yml -e username=$user -e password=$pass -e admin=yes --tags=register-user
```

### Serve the base domain as well (unless you already have something serving it)

```yaml
matrix_nginx_proxy_base_domain_serving_enabled: true

> ```

Reviewer 3: Maybe you should mention that each such change requires that you'd better go through `--tags=setup-all,start` again?

Well, if someone has read the docs or is familiar with Ansible, they might realize that.

### Disable all extras because of RAM

```yaml

> matrix_dimension_enabled: false

Reviewer 3: Dimension is disabled by default, because:
- it's not very useful, as mobile clients do not (yet) support custom integration managers.. and it doesn't have a responsive UI
- installing it is a bit involved (need to create a user and get an access token)
